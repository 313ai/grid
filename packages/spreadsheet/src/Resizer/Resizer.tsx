import React, { useRef, useEffect, useCallback } from 'react'

interface ResizerProps {
  onDrag?: (value: number) => void;
  initialTop: number;
  minTop?: number
}

const Resizer = ({ onDrag, initialTop, minTop = 0 }: ResizerProps) => {
  const dragging = useRef(false)
  const startY = useRef(0)
  const el = useRef<HTMLDivElement>(null)

  const handleMouseMove = useCallback((e: globalThis.MouseEvent) => {
    if (!el.current) {
      return
    }
    const diff = Math.max(e.pageY - startY.current, minTop)
    onDrag?.(diff)
    el.current.style.top = diff + 'px'
  }, [])
  const handleMouseDown = useCallback((e: React.MouseEvent<HTMLDivElement>) => {
    dragging.current = true
    var top = el.current?.style.top ? parseInt(el.current?.style.top) : 0
    startY.current = e.pageY - top;
    document.addEventListener('mousemove', handleMouseMove)
    document.addEventListener('mouseup', handleMouseUp)
  }, [])
  const handleMouseUp = useCallback(() => {
    dragging.current = false
    document.removeEventListener('mousemove', handleMouseMove)
    document.removeEventListener('mouseup', handleMouseUp)
  }, [])
  
  return (
    <div
      onMouseDown={handleMouseDown}
      ref={el}
      style={{
        height: 4,
        position: 'absolute',
        left: 0,
        top: initialTop,
        right: 0,
        zIndex: 1,
        cursor: 'ns-resize'
      }}
    />
  )
}

export default Resizer
