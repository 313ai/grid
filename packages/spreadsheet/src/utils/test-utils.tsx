import React from "react";
import { render } from "@testing-library/react";
import { Stage, Layer } from "react-konva";
import { ThemeProvider, theme } from "@chakra-ui/core";
import Konva from "konva";

const KonvaWrapper = ({ children }) => {
  return (
    <Stage>
      <ThemeWrapper>
        <Layer>{children}</Layer>
      </ThemeWrapper>
    </Stage>
  );
};

const konvaRenderer = (ui) => render(ui, { wrapper: KonvaWrapper });
const domRenderer = (ui) => render(ui, { wrapper: ThemeWrapper });

export * from "@testing-library/react";
export { konvaRenderer, domRenderer };

export const ThemeWrapper = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
