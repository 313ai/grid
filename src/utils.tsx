import React from "react";
import { Line, Rect } from "react-konva";
import { ShapeConfig } from "konva/types/Shape";

interface BoxProps extends ShapeConfig {
  onCrosshairMouseDown?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseMove?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseUp?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Create a box with custom top/right/bottom/left colors and widths
 * @param param0
 */
export const createBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
}: BoxProps) => {
  const commonProps = {
    perfectDrawEnabled: false,
    shadowForStrokeEnabled: false,
    hitStrokeWidth: 0,
    listening: false,
    lineCap: "square",
  };
  const lines = [
    <Line
      points={[x, y, x + width, y]}
      stroke={strokeTopColor}
      strokeWidth={strokeTopWidth}
      key="top"
      {...commonProps}
    />,
    <Line
      points={[x + width, y, x + width, y + height]}
      stroke={strokeRightColor}
      strokeWidth={strokeRightWidth}
      key="right"
      {...commonProps}
    />,
    <Line
      points={[x + width, y + height, x, y + height]}
      stroke={strokeBottomColor}
      strokeWidth={strokeBottomWidth}
      key="bottom"
      {...commonProps}
    />,
    <Line
      points={[x, y + height, x, y]}
      stroke={strokeLeftColor}
      strokeWidth={strokeLeftWidth}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <Rect
          fill={fill}
          x={x}
          y={y}
          width={width}
          height={height}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};

export const FillHandle: React.FC<BoxProps> = ({
  x = 0,
  y = 0,
  stroke,
  onMouseDown,
}) => {
  if (x === 0 || y === 0) return null;
  const crossHairSize = 8;
  const crossHairBorderSize = 1;
  return (
    <div
      style={{
        position: "absolute",
        left: x - crossHairSize / 2,
        top: y - crossHairSize / 2,
        width: crossHairSize,
        height: crossHairSize,
        border: `${crossHairBorderSize}px white solid`,
        background: stroke,
        cursor: "crosshair",
      }}
      onMouseDown={onMouseDown}
    />
  );
};

export const createHTMLBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
  strokeStyle = "solid",
}: BoxProps) => {
  const commonProps = {};
  width = width - Math.floor(strokeWidth);
  y = y - Math.ceil(strokeWidth / 2);
  const lines = [
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: width,
        height: strokeTopWidth,
        borderWidth: 0,
        borderColor: strokeTopColor,
        borderTopWidth: strokeTopWidth,
        borderStyle: strokeStyle,
      }}
      key="top"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x + width,
        top: y,
        width: strokeRightWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeRightColor,
        borderRightWidth: strokeRightWidth,
        borderStyle: strokeStyle,
      }}
      key="right"
      {...commonProps}
    />,
    ,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y + height,
        width: width + strokeTopWidth,
        height: strokeBottomWidth,
        borderWidth: 0,
        borderColor: strokeBottomColor,
        borderBottomWidth: strokeBottomWidth,
        borderStyle: strokeStyle,
      }}
      key="bottom"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: strokeLeftWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeLeftColor,
        borderLeftWidth: strokeLeftWidth,
        borderStyle: strokeStyle,
      }}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <div
          style={{
            position: "absolute",
            top: y,
            left: x,
            height,
            width,
            backgroundColor: fill,
            userSelect: "none",
            pointerEvents: "none",
          }}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};
