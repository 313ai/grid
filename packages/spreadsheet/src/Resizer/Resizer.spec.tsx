import React, { useState, useRef } from "react";
import { domRenderer, cleanup, fireEvent, act } from "./../utils/test-utils";
import Resizer from "./Resizer";

global.document.execCommand = jest.fn();

describe("Resizer", () => {
  afterEach(cleanup);
  it("renders spreadsheet", () => {
    const renderGrid = () => domRenderer(<Resizer initialTop={30} />);
    expect(renderGrid).not.toThrow();
  });
  it("matches snapshot", () => {
    const { asFragment } = domRenderer(<Resizer initialTop={30} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
