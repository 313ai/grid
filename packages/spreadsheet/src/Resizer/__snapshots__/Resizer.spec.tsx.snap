// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resizer matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="user-select: none; height: 4px; position: absolute; left: 0px; margin-top: -4px; right: 0px; z-index: 1; cursor: ns-resize;"
  />
</DocumentFragment>
`;
